Add complete

Aflynn50 opened the pull request:
This PR adds the complete stage. In this stage, we process any query arguments that are needed for execution. All structs mentioned in input expressions are passed to complete and the values that we need are extracted from them.

TheSignPainter98 commented on internal/expr/complete.go:
```diff
@@ -0,0 +1,49 @@
+
+	var tv = make(typeNameToValue)
+	for _, arg := range args {
+		if arg == (any)(nil) {
```
Is the `(any)` cast needed here?

Aflynn50 replied:
It doesn't seem like it, I copied this snippet from the reflection bit, not sure if its needed there either actually.

TheSignPainter98 commented on internal/expr/complete.go:
```diff
@@ -0,0 +1,49 @@
+		tv[reflect.TypeOf(arg).Name()] = arg
+	}
+
+	// Query parameteres.
```
Minor typo `s/parameteres/parameters/`

Aflynn50 replied:
Changed

TheSignPainter98 commented on internal/expr/complete.go:
```diff
@@ -0,0 +1,49 @@
+	var tv = make(typeNameToValue)
+	for _, arg := range args {
+		if arg == (any)(nil) {
+			return nil, fmt.Errorf("cannot reflect nil value")
```
Somehow 'reflect' seems too implementation-specific, perhaps `"got a nil argument"`?

Aflynn50 replied:
Changed

TheSignPainter98 commented on internal/expr/prepare.go:
```diff
@@ -0,0 +1,185 @@
+// Prepare takes a parsed expression and struct instantiations of all the type
+// mentioned in it.
+// The IO parts of the statement are checked for validity against the types
+// and expanded if necessary.
```
Super two minor things---the last word on the first line should be 'types' and I think the line-break between the two sentences could be removed.

I'm not sure whether this is the PR to fix these, though,

Aflynn50 replied:
True, they could look a lot better, I could fix it in the other PR.

Aflynn50 replied:
Ah, just checked and it has already been fixed in the other PR. I guess the merge got confused or something.

TheSignPainter98 commented on internal/expr/complete.go:
```diff
@@ -0,0 +1,49 @@
+
+	for _, p := range pe.ParsedExpr.queryParts {
+		if p, ok := p.(*inputPart); ok {
+			v, ok := tv[p.source.prefix]
```
I realise this probably isn't something to solve here, but somehow the `prefix` field storing the name of a type feels odd. Perhaps calling this `typeName` would be more natural?

Aflynn50 replied:
Yea, I'm not happy with the names of these either. They were decided on at the start after a lot of back and forth though, so I think we have to wait until that's a distant memory before changing them again.

TheSignPainter98 commented on internal/expr/complete.go:
```diff
@@ -0,0 +1,49 @@
+
+type typeNameToValue = map[string]any
+
+// Complete extracts query arguments specified in inputParts from structs.
```
I misunderstood this as `inputParts` being extracted from structs, I think the terms need a little re-ordering, perhaps something like
```go
// Complete populates query arguments
```

Also the name 'Complete' seems a little vague to me? Perhaps it could be something like 'Populate' or 'Args'?

Aflynn50 replied:
Changed it make it more clear

Aflynn50 committed:
# This is a combination of 7 commits.
# This is the 1st commit message:

First rebase commit

# The commit message #2 will be skipped:

# Third rebase commit

# The commit message #3 will be skipped:

# Third rebase commit

# The commit message #4 will be skipped:

# Third rebase commit

# The commit message #5 will be skipped:

# Third rebase commit

# The commit message #6 will be skipped:

# parent b025652aeddf9e4996fec0c775d704f39530ba1a
# author Alastair Flynn <alastairflynn@hotmail.co.uk> 1676306192 -0500
# committer Alastair Flynn <alastairflynn@hotmail.co.uk> 1676306192 -0500
# gpgsig -----BEGIN PGP SIGNATURE-----
#
#  iQGzBAABCgAdFiEEUE33GHeSvat7XeevMUkxPL/evRgFAmPqZxEACgkQMUkxPL/e
#  vRiU2Qv+Pg1juqMSmA9dxQx4M4VLBX5RlAB0l8rdYPG4BH99UT7anWWhOCVTgf0N
#  lvEivQYUnJj19BslGrzDF0vQUC1IEw6YRstBNb5pJtGsp8U99I888BSV9yvAj6k0
#  bXyOOtmtulfUgTqrwG9qVZoZq9wOoo5pxrMY5YcPyF0HizkcVDYi7FUfl04sFFLr
#  bnQ0pOdnRQTbvorHzraIue76H2W+stnXqEcjacqip73bHeiy78xa4hDBd8Q96TcM
#  vXxMp5bU8GSmTW5Ypihh3jeH3tFo5btP7AuknOcScw54LbhMXiNsfJeeklhXoB3s
#  OHUE6JbX2HAWFP1xQMxPkqq3+UqUGyAz/Wpa+o9UMCDs7oEoOl/zOGCR9bSz6Vkl
#  TTXieAeDr2Oukpv2tM+erDVEQdxZBJb2v2X05Ajn2n7vqRvl4valXtp20G9cKead
#  owfs6onGJj0BiMZkW2wDvQTkePQi6rbbd3AVJHt5rephIl9YXqSBRKRVmlUdSnVQ
#  ZSbadQRZ
#  =pc/K
#  -----END PGP SIGNATURE-----
#
# Third rebase commit

# The commit message #7 will be skipped:

# fixup after messy rebase

Aflynn50 committed:
Add complete expr

Aflynn50 committed:
Change test error messages

Aflynn50 committed:
Change error message to use .Type().String() instead of .Type().Name()
    for more accuracy

Aflynn50 committed:
Change type in error message to use Name() not String()

Aflynn50 committed:
Add comment

niemeyer commented on internal/expr/complete.go:
```diff
@@ -0,0 +1,52 @@
+type CompletedExpr struct {
+	outputs []loc
+	SQL     string
+	Args    []any
```
Per conversation a couple of weeks ago, let's please keep the details of the expression internal for now. For testing we can have a method or a function in export_test.go.

Aflynn50 replied:
Changed.

Aflynn50 committed:
Make args private and add export_test.go

niemeyer commented on internal/expr/complete.go:
```diff
@@ -0,0 +1,52 @@
+
+// Complete gathers the arguments that are specified in the query input parts from
+// structs passed as parameters.
+func (pe *PreparedExpr) Complete(args ...any) (ce *CompletedExpr, err error) {
```
The doc seems oriented a bit towards the internal implementation instead of the user of the function. Perhaps:
```
// Complete returns a completed expression ready for execution, using the provided values to
// substitute the input placeholders in the prepared expression. These placeholders use
// the syntax "$Person.fullname", where Person would be a type such as:
//
//        type Person struct {
//                Name string `db:"fullname"`
//        }
```

Aflynn50 replied:
Changed.

niemeyer commented on internal/expr/complete.go:
```diff
@@ -0,0 +1,52 @@
+func (pe *PreparedExpr) Complete(args ...any) (ce *CompletedExpr, err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("parameter issue: %s", err)
```
Maybe: "invalid input parameters: %s"

Aflynn50 replied:
Changed.

niemeyer commented on internal/expr/complete.go:
```diff
@@ -0,0 +1,52 @@
+	}()
+
+	var tv = make(map[reflect.Type]reflect.Value)
+	var typeNames []string
```
Eitehr tv and tn, or typeValue and typeNames?   Latter seems better. :)

Aflynn50 replied:
Changed to the latter.

niemeyer commented on internal/expr/complete.go:
```diff
@@ -0,0 +1,52 @@
+	var typeNames []string
+	for _, arg := range args {
+		if arg == nil {
+			return nil, fmt.Errorf("nil parameter")
```
Maybe: "need valid struct values, got nil"

Aflynn50 replied:
Changed.

niemeyer commented on internal/expr/expr_test.go:
```diff
@@ -565,3 +566,95 @@ func (s *ExprSuite) TestPrepareMismatchedColsAndTargs(c *C) {
+		c.Fatal(err)
+	}
+	_, err = preparedExpr.Complete(shadowedP)
+	c.Assert(err, ErrorMatches, `parameter issue: type Person not found, have: Person`)
```
Instead of the confusing error message, we might special case this situation and if the name is already missing (in other words, we're definitely in an error case), we might iterate over the set of types to see if there is another type with the same name. If so, we might special case the error message along the lines of:
```
"invalid input parameters: expected type foobar.Type but got fooser.Type"
```

Aflynn50 replied:
Added special case.

niemeyer commented on internal/expr/prepare.go:
```diff
@@ -11,14 +11,14 @@ import (
-	inputs  []*inputPart
+	outputs []loc
+	inputs  []loc
 	SQL     string
```
I've overlooked this on previous PRs. I believe we also want this as an opaque type.

Aflynn50 replied:
Changed and function in `export_test.go` added.

niemeyer commented on internal/expr/prepare.go:
```diff
@@ -11,14 +11,14 @@ import (
-	structType reflect.Type
-	field      field
+type loc struct {
+	typ   reflect.Type
```
The previous names seemed more clear and descriptive, even more for a type that is global to the package. Maybe the original meaning must change because you want it more generic, but would be nice to have a name that defines a bit better when we shoud expect to see this and for what purpose.

As we discussed, I think we can use just the `field` value itself here, as long as we have a backreference to its type.

Aflynn50 replied:
Changed to use just `field`s and to store the struct type in the `field` struct defined in `info.go`

niemeyer reviewed (changes requested):
Last pass, per our conversation. Looking good!

Aflynn50 committed:
Edit error messages

Aflynn50 committed:
Change clashing name error message

Aflynn50 committed:
Made the SQL field of PreparedExpr private

Aflynn50 committed:
Swap loc for field and store the struct type in the field

niemeyer reviewed (approved):
Thank you!
